### PR TITLE
book: chapter 4 part 1

### PR DESCRIPTION
**Pull request by frankfarzan**
Part of #1573

Also:
- add a place-holder overview page #1681
- minor clean ups

### CODE REVIEW

File: site/book/02-concepts/02-workflows.md [hunk 1]
```diff
@@ -18,14 +18,13 @@ For example, when consuming an upstream package, the initial workflow can look l
...
-rendering ensures that the `Kptfile` is valid.
+cluster. Render is a required step as it ensures certain preconditions and postconditions hold
+true about the state of the package.
 
-This workflow is an iterative process. There is usually a tight Edit<->Render loop in order to
+This workflow is an iterative process. There is usually a tight Edit-Render loop in order to
 produce the desired outcome.
 
-Some time later, you may want to update to a newer version of the upstream package:
+Now, some time later, you may want to update to a newer version of the upstream package:
```
Comment (mikebz) [inline site/book/02-concepts/02-workflows.md:27]: not blocking: not sure if this is better than before.

File: site/book/03-packages/05-updating-a-package.md [hunk 2]
```diff
@@ -2,6 +2,24 @@ An independent package records the exact commit where the local fork and the ups
...
+First, to see the changes you've made to the fork of the upstream package:
+
+```shell
+$ git diff
+```
+
+If you're happy with the changes, commit them:
+
+```shell
+$ git add .; git commit -m "My changes"
```
Comment (mikebz) [inline site/book/03-packages/05-updating-a-package.md:18]: this is probably a thing for a different PR but I found that the git command need to be executed in the folder and the kpt commands are assumed to be executing from a parent directory.  I thinking any mid level engineer is going to pick up on that very quickly so it's not a blocking thing, but might to address that somehow later.

File: site/book/04-using-functions/01-declarative-function-execution.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+2. Similarly, executes all the validators declared in the `mysql` package. The input to the first
+   validator is the output of the last mutator. The output of the last validator is the
+   output of the pipeline in the `mysql` package.
+3. Sequentially executes the list of mutators declared in the `wordpress` package. The input to the
+   first function is the union of:
+
+   - Resources read from configuration files in the `wordpress` package AND
+   - Output of the pipeline from the `mysql` package (Step 2).
+
+     Note that it is possible for a mutator in `wordpress` to modify resources in the `mysql` subpackage.
```
Comment (mikebz) [inline site/book/04-using-functions/01-declarative-function-execution.md:93]: NB: possible to outline here that the parent/child relationship is the reason.  It's implied but might be worth reiterating.

File: site/book/02-concepts/03-functions.md [hunk 4]
```diff
@@ -44,7 +44,8 @@ where:
 
 - `input items`: The input list of KRM resources to operate on.
 - `output items`: The output list obtained from adding, removing, or modifying items in the input.
-- `functionConfig`: An optional meta resource used to parameterize this invocation of the function.
+- `functionConfig`: An optional meta resource containing the arguments to this invocation of the
+  function.
```
[the remark this answers is not in the document]
Comment (frankfarzan) [reply site/book/02-concepts/03-functions.md:48]: Defined in the concepts chapters: https://kpt.dev/book/02-concepts/01-packages

File: site/book/03-packages/04-rendering-a-package.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+`render` is a critical step in the package lifecycle. At a high level,
+it perform the following steps:
+
+1. Enforces package preconditions. For example, it validates the `Kptfile`.
+2. Executes functions declared in the package hierarchy in a depth-first order. By default, the
+   packages are modified in-place.
+3. Guarantees package postconditions. For example, it enforces a consistent formatting of resources,
+   even though a function (developed by different people using different toolchains) may have modified the formatting in some way.
+
```
[the remark this answers is not in the document]
Comment (frankfarzan) [reply site/book/03-packages/04-rendering-a-package.md:17]: Discussed offline.

File: site/book/03-packages/05-updating-a-package.md (hunk 2 above)
Comment (frankfarzan) [reply site/book/03-packages/05-updating-a-package.md:18]: We should clarify what the current working directory is when invoking commands. I try to be consistent if sequentially executing all commands in the book, but we can't really rely on a `cd` from few pages back.

Open to ideas to follow up here:
1) Annotate each block using a comment, e.g.:

```shell
// From wordpress/
$ kpt
```

2) always `cd` into a dir (not idemotent)

File: site/book/04-using-functions/01-declarative-function-execution.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+4. Similarly, execute all the validators declared in the `wordpress` package. The output of the last
+   validator is the output of the pipeline in the `wordpress` package.
+5. Write the output of step 4 by modifying the local filesystem in-place. This can change both
+   `wordpress` and `mysql` packages.
```
[the remark this answers is not in the document]
Comment (frankfarzan) [reply site/book/04-using-functions/01-declarative-function-execution.md:98]: Yes, this domain lends itself well to visualization. Let's follow up.

**Pull request merged**